File: backend/Tasque.BLL/Services/SprintService.cs
using AutoMapper;
using FluentValidation;
using Microsoft.EntityFrameworkCore;
using Tasque.Core.BLL.Exeptions;
﻿using FluentValidation;
using Tasque.Core.Common.DTO.Sprint;
using Tasque.Core.Common.DTO.Task;
using Tasque.Core.Common.DTO.User;
using Tasque.Core.Common.Entities;
using Tasque.Core.DAL;
using Task = System.Threading.Tasks.Task;
using Tasque.Core.Common.Models.Task;

namespace Tasque.Core.BLL.Services
{
    public class SprintService : EntityCrudService<Sprint>
    {
        private IMapper _mapper;
        public SprintService(DataContext db, IMapper mapper) : base(db)
        {
            _mapper = mapper;
        }

        public async Task<IEnumerable<SprintDto>> GetProjectSprints(int projectId)
        {
            var sprints = await _db.Sprints
                .Where(s => s.ProjectId == projectId)
                .ToListAsync();

            return _mapper.Map<IEnumerable<SprintDto>>(sprints);
        }
        public async Task<IEnumerable<TaskDto>> GetSprintTasks(int sprintId)
        {
            var tasks = await _db.Tasks
                .Where(t => t.SprintId == sprintId)
                .ToListAsync();

            return _mapper.Map<IEnumerable<TaskDto>>(tasks);
        }

        public async Task<IEnumerable<UserDto>> GetSprintUsers(int sprintId)
        {
            var users = await _db.Tasks
                .Include(t => t.Author)
                .Where(t => t.SprintId == sprintId)
                .Select(t =>t.Author)
                .GroupBy(x => x.Id)
                .Select(x => x.First())
                .ToListAsync();

            return _mapper.Map<IEnumerable<UserDto>>(users);
        }

        public async Task<Sprint> Edit(EditSprintDto dto)
        {
            var entity = await _db.Sprints.FirstOrDefaultAsync(s => s.Id == dto.Id)
                ?? throw new ValidationException("Sprint not found");
            entity.Name = dto.Name;
            if (dto.StartAt != null)
            {
                entity.StartAt = DateTime.SpecifyKind((DateTime)dto.StartAt, DateTimeKind.Utc);
            }
            if (dto.EndAt != null)
            {
                entity.EndAt = DateTime.SpecifyKind((DateTime)dto.EndAt, DateTimeKind.Utc);
            }
            entity.Description = dto.Description;
            entity.UpdatedAt = DateTime.UtcNow;
            
            if (dto.IsStarting && dto.Tasks != null)
            {
                foreach(var taskId in dto.Tasks)
                {
                    var task = await _db.Tasks.FirstOrDefaultAsync(t => t.Id == taskId)
                        ?? throw new ValidationException("Task not found");
                    task.SprintId = dto.Id;
                    _db.Tasks.Update(task);
                }
            }
            
            _db.Sprints.Update(entity);
            await _db.SaveChangesAsync();
            return entity;
        }

        public async Task CompleteSprint(int sprintId)
        {
            var sprint = await _db.Sprints.FirstOrDefaultAsync(s => s.Id == sprintId);

            if (sprint == null)
                throw new HttpException(System.Net.HttpStatusCode.NotFound, "Sprinter with this ID does not exist");

            sprint.IsComplete = true;

            _db.Update(sprint);
            await _db.SaveChangesAsync();
        }

        public async Task UpdateTaskEstimate(TaskEstimateUpdate taskEstimateUpdate)
        {
            var sprint = await _db.Sprints
                .FirstOrDefaultAsync(s => s.Id == taskEstimateUpdate.SprintId);

            if (sprint == null)
                throw new HttpException(System.Net.HttpStatusCode.NotFound, "Sprinter with this ID does not exist");

            var task = await _db.Tasks
                .FirstOrDefaultAsync(t => t.Id == taskEstimateUpdate.TaskId);

            if (task == null)
                throw new HttpException(System.Net.HttpStatusCode.NotFound, "Task with this ID does not exist");

            task.Estimate = taskEstimateUpdate.Estimate;

            _db.Update(task);
            await _db.SaveChangesAsync();
        }
    }
}

File: frontend/src/core/services/sprint.service.ts
import { HttpResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { TaskModel } from '../models/task/task-model';
import { UserModel } from '../models/user/user-model';
import { Observable } from 'rxjs';
import { EditSprintModel } from '../models/sprint/edit-sprint-model';
import { SprintModel } from '../models/sprint/sprint-model';
import { HttpService } from './http.service';
import { TaskEstimateUpdate } from '../models/task/task-estimate-update';
import { TaskModelDto } from '../models/task/task-model-dto';

@Injectable({
  providedIn: 'root',
})
export class SprintService {
  public routePrefix = '/api/sprint';

  constructor(public httpService: HttpService) {}

  completeSprint(sprintId: number): Observable<void> {
    return this.httpService.putRequest<void>(
      this.routePrefix + `/complete/${sprintId}`,
      new Object(),
    );
  }

  getProjectSprints(
    projectId: number,
  ): Observable<HttpResponse<SprintModel[]>> {
    return this.httpService.getFullRequest<SprintModel[]>(
      this.routePrefix + `/getSprintsByProjectId/${projectId}`,
    );
  }

  getSprintTasks(sprintId: number): Observable<HttpResponse<TaskModelDto[]>> {
    return this.httpService.getFullRequest<TaskModelDto[]>(
      this.routePrefix + `/${sprintId}/tasks`,
    );
  }

  getSprintUsers(sprintId: number): Observable<HttpResponse<UserModel[]>> {
    return this.httpService.getFullRequest<UserModel[]>(
      this.routePrefix + `/${sprintId}/users`,
    );
  }

  editSprint(
    editedSprint: EditSprintModel,
  ): Observable<HttpResponse<SprintModel>> {
    return this.httpService.putFullRequest<SprintModel>(
      this.routePrefix + '/edit',
      editedSprint,
    );
  }

  updareSprint(
    sprintId: number,
    editedSprint: SprintModel,
  ): Observable<HttpResponse<SprintModel>> {
    return this.httpService.putFullRequest<SprintModel>(
      this.routePrefix + `/update/${sprintId}`,
      editedSprint,
    );
  }

  updateTaskEstimate(
    taskEstimateUpdate: TaskEstimateUpdate,
  ): Observable<HttpResponse<void>> {
    return this.httpService.putFullRequest<void>(
      this.routePrefix + '/estimate',
      taskEstimateUpdate,
    );
  }
}
